vlasovsolver/gpu_acc_map.cpp

```diff
@@ -1074,34 +1126,45 @@ __global__ void __launch_bounds__(WID3,Hashinator::defaults::MAX_BLOCKSIZE/(WID3
 
                // compute right integrand
                #ifdef ACC_SEMILAG_PLM
-               target_density_r = v_norm_r * ( a[0] + v_norm_r * a[1] );
+               target_density_r = a[1];
+               target_density_r = a[0] + v_norm_r * target_density_r;
+               target_density_r = v_norm_r * target_density_r;
                #endif
                #ifdef ACC_SEMILAG_PPM
-               target_density_r = v_norm_r * ( a[0] + v_norm_r * ( a[1] + v_norm_r * a[2] ) );
+               target_density_r = a[2];
+               target_density_r = a[1] + v_norm_r * target_density_r;
+               target_density_r = a[0] + v_norm_r * target_density_r;
+               target_density_r = v_norm_r * target_density_r;
                #endif
                #ifdef ACC_SEMILAG_PQM
-               target_density_r = v_norm_r * ( a[0] + v_norm_r * ( a[1] + v_norm_r * ( a[2] + v_norm_r * ( a[3] + v_norm_r * a[4] ) ) ) );
+               target_density_r = a[4];
+               target_density_r = a[3] + v_norm_r * target_density_r;
+               target_density_r = a[2] + v_norm_r * target_density_r;
+               target_density_r = a[1] + v_norm_r * target_density_r;
+               target_density_r = a[0] + v_norm_r * target_density_r;
+               target_density_r = v_norm_r * target_density_r;
+
+               //target_density_r = v_norm_r * ( a[0] + v_norm_r * ( a[1] + v_norm_r * ( a[2] + v_norm_r * ( a[3] + v_norm_r * a[4] ) ) ) );
                #endif
 
                // integral area between the two integrands
                Realf tval = target_density_r - target_density_l;
 
                // Store directly into adjusted velocity block container at correct target GID/LID
-               const vmesh::GlobalID targetGID =
-                  col_i  * gpu_block_indices_to_id[0] +
-                  col_j  * gpu_block_indices_to_id[1] +
-                  blockK * gpu_block_indices_to_id[2];
+               const vmesh::GlobalID targetGID = columnGID + blockK * gpu_block_indices_to_id[2];
                const vmesh::LocalID targetLID = vmesh->getLocalID(targetGID);
                // The target velocity cell within the target bloxk
                const int tcell = target_cell_index_common
                                + gk_mod_WID * gpu_cell_indices_to_id[2];
                // Write values into block data
-               if (isfinite(tval) && (tval>0) && (targetLID != invalidLID) ) {
-                  gpu_blockData[targetLID * WID3 + tcell] += tval;
-                  // atomicAdd(&gpu_blockData[targetLID*WID3+tcell],tval);
+               if (isfinite(tval) && (tval>(Realf)(0.0)) && (targetLID != invalidLID) ) {
+                  // gpu_blockData[targetLID * WID3 + tcell] += tval;
+
+                  // We use atomicAdd to avoid the need for sync
+                  // It shouldn't be any slower if there is no competition
+                  atomicAdd(&gpu_blockData[targetLID*WID3+tcell],tval);
```

Comment:
This should be also checked on AMD hardware, whether it's the case! Floating point atomic operations on AMD can behave *very* differently.... (insert can of worms)